File: backend/app/api/v1/endpoints/site_settings.py
from typing import Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel, EmailStr

from app.api.deps import get_current_active_superuser
from app.models import SiteSetting, DEFAULT_SETTINGS, User
from app.schemas.site_setting import (
    SiteSettingResponse,
    SiteSettingUpdate,
    SiteSettingBulkUpdate,
    SiteSettingsResponse,
    PublicSiteSettingsResponse,
)
from app.schemas.response import Response, ResponseCode, BusinessError, success
from app.core.email import send_email

router = APIRouter()


class TestEmailRequest(BaseModel):
    """测试邮件请求"""
    email: EmailStr


@router.get("/public", response_model=Response[PublicSiteSettingsResponse])
async def get_public_settings():
    """Get public site settings (no authentication required)"""
    settings = await SiteSetting.get_all_by_category(public_only=True)
    return success(data=PublicSiteSettingsResponse(
        site_name=settings.get("site_name", "Clouisle"),
        site_description=settings.get("site_description", ""),
        site_url=settings.get("site_url", ""),
        site_icon=settings.get("site_icon", ""),
        allow_registration=settings.get("allow_registration", True),
        require_approval=settings.get("require_approval", False),
        email_verification=settings.get("email_verification", True),
        enable_captcha=settings.get("enable_captcha", False),
        allow_account_deletion=settings.get("allow_account_deletion", True),
    ))


@router.get("", response_model=Response[SiteSettingsResponse])
async def get_all_settings(
    category: Optional[str] = None,
    current_user: User = Depends(get_current_active_superuser),
):
    """Get all site settings (admin only)"""
    settings = await SiteSetting.get_all_by_category(category=category)
    return success(data=SiteSettingsResponse(settings=settings))


@router.get("/{key}", response_model=Response[SiteSettingResponse])
async def get_setting(
    key: str,
    current_user: User = Depends(get_current_active_superuser),
):
    """Get a specific setting by key (admin only)"""
    setting = await SiteSetting.filter(key=key).first()
    if not setting:
        raise BusinessError(
            code=ResponseCode.NOT_FOUND,
            msg=f"Setting '{key}' not found",
        )
    return success(data=SiteSettingResponse(
        key=setting.key,
        value=SiteSetting._convert_value(setting.value, setting.value_type),
        value_type=setting.value_type,
        category=setting.category,
        description=setting.description,
        is_public=setting.is_public,
    ))


@router.put("/{key}", response_model=Response[SiteSettingResponse])
async def update_setting(
    key: str,
    data: SiteSettingUpdate,
    current_user: User = Depends(get_current_active_superuser),
):
    """Update a specific setting (admin only)"""
    setting = await SiteSetting.filter(key=key).first()
    if not setting:
        # Check if it's a known default setting
        if key in DEFAULT_SETTINGS:
            config = DEFAULT_SETTINGS[key]
            setting = await SiteSetting.set_value(
                key=key,
                value=data.value,
                value_type=config["type"],
                category=config["category"],
                description=config["desc"],
                is_public=config["public"],
            )
        else:
            raise BusinessError(
                code=ResponseCode.NOT_FOUND,
                msg=f"Setting '{key}' not found",
            )
    else:
        setting = await SiteSetting.set_value(
            key=key,
            value=data.value,
            value_type=setting.value_type,
            category=setting.category,
            description=setting.description,
            is_public=setting.is_public,
        )
    
    return success(data=SiteSettingResponse(
        key=setting.key,
        value=SiteSetting._convert_value(setting.value, setting.value_type),
        value_type=setting.value_type,
        category=setting.category,
        description=setting.description,
        is_public=setting.is_public,
    ))


@router.put("", response_model=Response[SiteSettingsResponse])
async def bulk_update_settings(
    data: SiteSettingBulkUpdate,
    current_user: User = Depends(get_current_active_superuser),
):
    """Bulk update multiple settings (admin only)"""
    for key, value in data.settings.items():
        setting = await SiteSetting.filter(key=key).first()
        if setting:
            await SiteSetting.set_value(
                key=key,
                value=value,
                value_type=setting.value_type,
                category=setting.category,
                description=setting.description,
                is_public=setting.is_public,
            )
        elif key in DEFAULT_SETTINGS:
            config = DEFAULT_SETTINGS[key]
            await SiteSetting.set_value(
                key=key,
                value=value,
                value_type=config["type"],
                category=config["category"],
                description=config["desc"],
                is_public=config["public"],
            )
    
    # Return all settings
    settings = await SiteSetting.get_all_by_category()
    return success(data=SiteSettingsResponse(settings=settings))


@router.post("/reset", response_model=Response[SiteSettingsResponse])
async def reset_settings(
    category: Optional[str] = None,
    current_user: User = Depends(get_current_active_superuser),
):
    """Reset settings to default values (admin only)"""
    for key, config in DEFAULT_SETTINGS.items():
        if category and config["category"] != category:
            continue
        await SiteSetting.set_value(
            key=key,
            value=config["value"],
            value_type=config["type"],
            category=config["category"],
            description=config["desc"],
            is_public=config["public"],
        )
    
    settings = await SiteSetting.get_all_by_category(category=category)
    return success(data=SiteSettingsResponse(settings=settings))


@router.post("/test-email", response_model=Response[None])
async def send_test_email(
    data: TestEmailRequest,
    current_user: User = Depends(get_current_active_superuser),
):
    """Send a test email to verify SMTP configuration (admin only)"""
    # Check if SMTP is enabled
    smtp_enabled = await SiteSetting.get_value("smtp_enabled", False)
    if not smtp_enabled:
        raise BusinessError(
            code=ResponseCode.EMAIL_SEND_FAILED,
            msg_key="smtp_not_configured",
        )
    
    site_name = await SiteSetting.get_value("site_name", "Clouisle")
    
    subject = f"【{site_name}】测试邮件 / Test Email"
    body_text = f"""这是一封来自 {site_name} 的测试邮件。

如果您收到了这封邮件，说明 SMTP 配置正确。

---

This is a test email from {site_name}.

If you received this email, your SMTP configuration is working correctly."""

    body_html = f"""
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
</head>
<body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; max-width: 600px; margin: 0 auto; padding: 20px;">
    <h2 style="color: #333;">✅ SMTP 配置测试成功</h2>
    <p>这是一封来自 <strong>{site_name}</strong> 的测试邮件。</p>
    <p style="color: #666;">如果您收到了这封邮件，说明 SMTP 配置正确。</p>
    <hr style="border: none; border-top: 1px solid #eee; margin: 20px 0;">
    <h2 style="color: #333;">✅ SMTP Configuration Test Successful</h2>
    <p>This is a test email from <strong>{site_name}</strong>.</p>
    <p style="color: #666;">If you received this email, your SMTP configuration is working correctly.</p>
</body>
</html>
"""
    
    result = await send_email(data.email, subject, body_text, body_html)
    
    if not result:
        raise BusinessError(
            code=ResponseCode.EMAIL_SEND_FAILED,
            msg_key="email_send_failed",
        )
    
    return success(msg_key="test_email_sent")

File: frontend/lib/api/site-settings.ts
import { api } from './client'

export interface SiteSetting {
  key: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value: any
  value_type: string
  category: string
  description?: string
  is_public: boolean
}

export interface SiteSettings {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  settings: Record<string, any>
}

export interface PublicSiteSettings {
  site_name: string
  site_description: string
  site_url: string
  site_icon: string
  allow_registration: boolean
  require_approval: boolean
  email_verification: boolean
  enable_captcha: boolean
  allow_account_deletion: boolean
}

export interface GeneralSettings {
  site_name: string
  site_description: string
  site_url: string
  site_icon: string
  allow_registration: boolean
  require_approval: boolean
  email_verification: boolean
  allow_account_deletion: boolean
}

export interface SecuritySettings {
  min_password_length: number
  require_uppercase: boolean
  require_number: boolean
  require_special_char: boolean
  session_timeout_days: number
  single_session: boolean
  max_login_attempts: number
  lockout_duration_minutes: number
  enable_captcha: boolean
}

export interface EmailSettings {
  smtp_enabled: boolean
  smtp_host: string
  smtp_port: number
  smtp_encryption: 'none' | 'ssl' | 'tls'
  smtp_username: string
  smtp_password: string
  email_from_name: string
  email_from_address: string
}

export const siteSettingsApi = {
  /**
   * Get public site settings (no auth required)
   */
  async getPublic(): Promise<PublicSiteSettings> {
    return api.get<PublicSiteSettings>('/site-settings/public')
  },

  /**
   * Get all settings (admin only)
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async getAll(category?: string): Promise<Record<string, any>> {
    const params = category ? `?category=${category}` : ''
    const res = await api.get<SiteSettings>(`/site-settings${params}`)
    return res.settings
  },

  /**
   * Get a specific setting
   */
  async get(key: string): Promise<SiteSetting> {
    return api.get<SiteSetting>(`/site-settings/${key}`)
  },

  /**
   * Update a specific setting
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async update(key: string, value: any): Promise<SiteSetting> {
    return api.put<SiteSetting>(`/site-settings/${key}`, { value })
  },

  /**
   * Bulk update multiple settings
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async bulkUpdate(settings: Record<string, any>): Promise<Record<string, any>> {
    const res = await api.put<SiteSettings>('/site-settings', { settings })
    return res.settings
  },

  /**
   * Reset settings to default values
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async reset(category?: string): Promise<Record<string, any>> {
    const params = category ? `?category=${category}` : ''
    const res = await api.post<SiteSettings>(`/site-settings/reset${params}`, null)
    return res.settings
  },

  /**
   * Get general settings
   */
  async getGeneral(): Promise<GeneralSettings> {
    const settings = await this.getAll('general')
    return {
      site_name: settings.site_name ?? 'Clouisle',
      site_description: settings.site_description ?? '',
      site_url: settings.site_url ?? '',
      site_icon: settings.site_icon ?? '',
      allow_registration: settings.allow_registration ?? true,
      require_approval: settings.require_approval ?? false,
      email_verification: settings.email_verification ?? true,
      allow_account_deletion: settings.allow_account_deletion ?? true,
    }
  },

  /**
   * Update general settings
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async updateGeneral(data: Partial<GeneralSettings>): Promise<Record<string, any>> {
    return this.bulkUpdate(data)
  },

  /**
   * Get security settings
   */
  async getSecurity(): Promise<SecuritySettings> {
    const settings = await this.getAll('security')
    return {
      min_password_length: settings.min_password_length ?? 8,
      require_uppercase: settings.require_uppercase ?? true,
      require_number: settings.require_number ?? true,
      require_special_char: settings.require_special_char ?? false,
      session_timeout_days: settings.session_timeout_days ?? 30,
      single_session: settings.single_session ?? false,
      max_login_attempts: settings.max_login_attempts ?? 5,
      lockout_duration_minutes: settings.lockout_duration_minutes ?? 15,
      enable_captcha: settings.enable_captcha ?? false,
    }
  },

  /**
   * Update security settings
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async updateSecurity(data: Partial<SecuritySettings>): Promise<Record<string, any>> {
    return this.bulkUpdate(data)
  },

  /**
   * Get email settings
   */
  async getEmail(): Promise<EmailSettings> {
    const settings = await this.getAll('email')
    return {
      smtp_enabled: settings.smtp_enabled ?? false,
      smtp_host: settings.smtp_host ?? '',
      smtp_port: settings.smtp_port ?? 587,
      smtp_encryption: settings.smtp_encryption ?? 'tls',
      smtp_username: settings.smtp_username ?? '',
      smtp_password: settings.smtp_password ?? '',
      email_from_name: settings.email_from_name ?? 'Clouisle',
      email_from_address: settings.email_from_address ?? '',
    }
  },

  /**
   * Update email settings
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  async updateEmail(data: Partial<EmailSettings>): Promise<Record<string, any>> {
    return this.bulkUpdate(data)
  },

  /**
   * Send test email
   */
  async sendTestEmail(email: string): Promise<void> {
    await api.post<null>('/site-settings/test-email', { email })
  },
}
